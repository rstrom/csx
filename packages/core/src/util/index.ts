/** return typed keys array - for gotchas see: https://stackoverflow.com/a/55012175/3056284 */
export function keys<O>(o: O) {
  return Object.keys(o) as (keyof O)[];
}
