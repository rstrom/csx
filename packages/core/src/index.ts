import { InterpolationWithTheme, jsx, SerializedStyles } from "@emotion/core";
import React from "react";
import { keys } from "./util";

export const csx = <
  E extends HTMLElement = HTMLDivElement,
  S extends string = ""
>(
  baseStyle: SerializedStyles | SerializedStyles[],
  conditionalsObj: Record<S, SerializedStyles> | {} = {},
  elementType: keyof React.ReactDOM = "div"
): React.FC<Partial<Record<S, boolean>> &
  React.DetailedHTMLProps<React.HTMLAttributes<E>, E> & {
    css?: InterpolationWithTheme<never>;
  }> => props => {
  const baseStyles = Array.isArray(baseStyle) ? baseStyle : [baseStyle];
  const activeConditionals = keys(conditionalsObj)
    .filter(k => !!props[k])
    .map(k => conditionalsObj[k]);
  return jsx(elementType, {
    ...props,
    css: [...baseStyles, ...activeConditionals, props.css]
  });
};
