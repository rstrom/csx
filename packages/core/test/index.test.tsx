import css from "@emotion/css";
import React from "react";
import { csx } from "../src/index";
import renderer from "react-test-renderer";
import serializer from "jest-emotion";

expect.addSnapshotSerializer(serializer);

describe("Styling a Button component", () => {
  it("works", () => {
    const Button = csx(
      css`
        color: red;
      `,
      {
        foo: css`
          color: green;
        `,
        bar: css`
          color: yellow;
        `
      },
      "button"
    );
    const tree = renderer.create(<Button foo bar={false} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
