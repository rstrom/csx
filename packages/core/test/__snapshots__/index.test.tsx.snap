// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styling a Button component works 1`] = `
.emotion-0 {
  color: red;
  color: green;
}

<button
  bar={false}
  className="emotion-0"
  foo={true}
/>
`;
